demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,134 +142,347 @@ typedef struct
-     * are set by the initializer. */
-    xCredentials.pRootCa = pxHostAddressData->pcCertificate;
-    xCredentials.rootCaSize = ( size_t ) pxHostAddressData->ulCertificateSize;
-    /* Disable SNI. */
-    xCredentials.disableSni = true;
-    /* ALPN is not needed. */
-    xCredentials.pAlpnProtos = NULL;
-
-    /* Set the server info. */
+    ServerInfo_t xServerInfo = { 0 };
+
+    SocketsConfig_t xSocketsConfig = { 0 };
+    TransportSocketStatus_t xNetworkStatus = TRANSPORT_SOCKET_STATUS_SUCCESS;
+    BackoffAlgorithmContext_t xReconnectParams;
+    BaseType_t xBackoffStatus = pdFALSE;
+    BaseType_t xMqttStatus = pdFALSE;
+
+    /* Set the credentials for establishing a TLS connection. */
+    /* Initializer server information. */
     xServerInfo.pHostName = pxHostAddressData->pcHostAddress;
+    xServerInfo.hostNameLength = strlen( pxHostAddressData->pcHostAddress );
     xServerInfo.port = clientcredentialMQTT_BROKER_PORT;
 
-    /* Set the members of the network info. */
-    xNetworkInfo.createNetworkConnection = true;
-    xNetworkInfo.u.setup.pNetworkServerInfo = &xServerInfo;
-    xNetworkInfo.u.setup.pNetworkCredentialInfo = &xCredentials;
-    xNetworkInfo.pNetworkInterface = pNetworkInterface;
-
-    /* Connect to the broker. */
-    xMqttConnectInfo.awsIotMqttMode = true;
-    xMqttConnectInfo.cleanSession = true;
-    xMqttConnectInfo.pClientIdentifier = ( const char * ) ( clientcredentialIOT_THING_NAME );
-    xMqttConnectInfo.clientIdentifierLength = ( uint16_t ) ( strlen( clientcredentialIOT_THING_NAME ) );
-    xMqttConnectInfo.keepAliveSeconds = ggdDEMO_KEEP_ALIVE_INTERVAL_SECONDS;
-
-    /* Call MQTT's CONNECT function. */
-    for( connectAttempt = 0; connectAttempt < ggdDEMO_NUM_TRIES; connectAttempt++ )
+    /* Configure credentials for TLS mutual authenticated session. */
+    xSocketsConfig.enableTls = true;
+    xSocketsConfig.pAlpnProtos = NULL;
+    xSocketsConfig.maxFragmentLength = 0;
+    xSocketsConfig.disableSni = true;
+    xSocketsConfig.pRootCa = pxHostAddressData->pcCertificate;
+    xSocketsConfig.rootCaSize = ( size_t ) pxHostAddressData->ulCertificateSize;
+    xSocketsConfig.sendTimeoutMs = TRANSPORT_SEND_RECV_TIMEOUT_MS;
+    xSocketsConfig.recvTimeoutMs = TRANSPORT_SEND_RECV_TIMEOUT_MS;
+
+    /* Initialize reconnect attempts and interval. */
+    BackoffAlgorithm_InitializeParams( &xReconnectParams,
+                                       RETRY_BACKOFF_BASE_MS,
+                                       RETRY_MAX_BACKOFF_DELAY_MS,
+                                       RETRY_MAX_ATTEMPTS );
+
+    /* Attempt to connect to MQTT broker. If connection fails, retry after
+     * a timeout. Timeout value will exponentially increase till maximum
+     * attempts are reached.
+     */
+    do
     {
-        if( connectAttempt > 0 )
+        /* Establish a TLS session with the MQTT broker. This example connects to
+         * the MQTT broker as specified in pxHostAddressData->pcHostAddress and
```

Comment:
We can be more explicit here for more helpful comments
```suggestion
         * the Greengrass Core as specified in pxHostAddressData->pcHostAddress and
```